Sources/GRPCCore/GRPCClient.swift

```diff
@@ -0,0 +1,823 @@
+      method: descriptor,
+      configuration: applicableConfiguration,
+      serializer: serializer,
+      deserializer: deserializer,
+      transport: transport,
+      interceptors: interceptors,
+      handler: handler
+    )
+  }
+
+  /// Start a bidirectional streaming RPC.
+  ///
+  /// - Note: ``run()`` must have been called and still executing, and ``close()`` mustn't
+  /// have been called.
+  ///
+  /// - Parameters:
+  ///   - request: The streaming request.
+  ///   - descriptor: The method descriptor for which to execute this request.
+  ///   - serializer: A request serializer.
+  ///   - deserializer: A response deserializer.
+  ///   - handler: A response stream handler.
+  ///
+  /// - Returns: The return value from the `handler`.
+  public func bidirectionalStreaming<Request, Response, ReturnValue>(
+    request: ClientRequest.Stream<Request>,
+    descriptor: MethodDescriptor,
+    serializer: some MessageSerializer<Request>,
+    deserializer: some MessageDeserializer<Response>,
+    handler: @Sendable @escaping (ClientResponse.Stream<Response>) async throws -> ReturnValue
+  ) async throws -> ReturnValue {
+    let (configurationOverrides, interceptors) = try self.storage.withLockedValue { storage in
+      switch storage.state {
+      case .running:
+        return (storage.methodConfigurationOverrides, storage.interceptors.values)
+      case .notStarted:
+        throw ClientError(
+          code: .clientIsNotRunning,
+          message: "Client must be running to make an RPC: call run() first."
+        )
+      case .stopping, .stopped:
+        throw ClientError(
+          code: .clientIsStopped,
+          message: "Client has been stopped. Can't make any more RPCs."
+        )
+      }
+    }
+
+    let applicableConfiguration = self.resolveMethodConfiguration(
+      descriptor: descriptor,
+      clientConfigurations: configurationOverrides
+    )
+
+    return try await ClientRPCExecutor.execute(
+      request: request,
+      method: descriptor,
+      configuration: applicableConfiguration,
+      serializer: serializer,
+      deserializer: deserializer,
+      transport: transport,
+      interceptors: interceptors,
+      handler: handler
+    )
+  }
+
+  private func resolveMethodConfiguration(
+    descriptor: MethodDescriptor,
+    clientConfigurations configurationOverrides: MethodConfigurationRegistry
+  ) -> MethodConfiguration {
+    if let clientOverride = configurationOverrides[descriptor, useDefault: false] {
+      return clientOverride
+    }
+
+    if let transportConfiguration = self.transport.executionConfiguration(forMethod: descriptor) {
+      return transportConfiguration
+    }
+
+    return configurationOverrides[descriptor]
+  }
+}
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+extension GRPCClient {
+  /// A collection of interceptors providing cross-cutting functionality to each accepted RPC.
+  public struct Interceptors: Sendable {
```

Comment:
Do we ever want to actually make this fully generic? So you can have a `GRPCCLient` with a generic interceptor pipeline which hopefully gets specialised. @glbrntt

Reply:
I don't think the generics carry their weight at the interceptor level tbh. Having a client generic over a transport and a bunch of interceptors hurts ergonomics quite badly.